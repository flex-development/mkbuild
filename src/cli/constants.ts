/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
