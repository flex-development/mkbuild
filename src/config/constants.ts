/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
