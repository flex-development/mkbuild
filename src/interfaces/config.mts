/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystem, Task } from '@flex-development/mkbuild'

/**
 * Build configuration.
 *
 * @see {@linkcode Task}
 *
 * @extends {Task}
 */
interface Config extends Task {
  /**
   * File system API.
   *
   * @see {@linkcode FileSystem}
   */
  fs?: Partial<FileSystem> | null | undefined

  /**
   * Build tasks.
   *
   * Each build task will be merged with the base task. Top-level lists (arrays
   * and sets) are concatted and deduped. Plain objects are deep merged.
   *
   * > ðŸ‘‰ **Note**: If empty, `null`, or `undefined`, a single build task will
   * > be inferred from the base task.
   *
   * @see {@linkcode Task}
   */
  tasks?: Task[] | null | undefined
}

export type { Config as default }
