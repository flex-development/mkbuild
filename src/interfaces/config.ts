/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystemAdapter } from '#src/types'
import type * as esbuild from 'esbuild'
import type Options from './options'
import type Task from './task'

/**
 * Build configuration options.
 *
 * @extends {Options}
 */
interface Config extends Options {
  /**
   * Load build config from config file.
   *
   * @default true
   */
  configfile?: boolean

  /**
   * Build entries.
   *
   * **Note**: If empty or `undefined`, a single build entry will be inferred
   * from the remaining set of configuration options.
   *
   * @default []
   */
  entries?: Partial<Omit<Task, 'write'>>[]

  /**
   * Custom implementations of file system methods.
   *
   * @see {@linkcode FileSystemAdapter}
   * @see https://nodejs.org/api/fs.html
   *
   * @default fsa
   */
  fs?: FileSystemAdapter

  /**
   * Serve files.
   *
   * @see {@linkcode esbuild.ServeOptions}
   * @see https://esbuild.github.io/api/#serve
   *
   * @default false
   */
  serve?: esbuild.ServeOptions | boolean

  /**
   * Watch files.
   *
   * @default false
   */
  watch?: boolean

  /**
   * Write build results.
   *
   * @default false
   */
  write?: boolean
}

export type { Config as default }
