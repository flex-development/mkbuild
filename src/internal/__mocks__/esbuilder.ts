/**
 * @file Mock Internals - esbuilder
 * @module mkbuild/internal/mocks/esbuilder
 */

/**
 * `esbuilder` module type.
 */
type Actual = typeof import('../esbuilder')

/**
 * `esbuilder` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('../esbuilder')

export default vi.fn(actual.default)
