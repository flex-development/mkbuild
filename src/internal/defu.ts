/**
 * @file Internal - defu
 * @module mkbuild/internal/defu
 */

import {
  isUndefined,
  type ExactOptionalPropertyTypes,
  type ObjectPlain
} from '@flex-development/tutils'
import { mergeAndCompare as merge, type Merge } from 'merge-anything'

/**
 * Assigns default properties to the given `object`.
 *
 * Leftmost arguments have more priority when assigning defaults.
 *
 * @template T - Destination object type
 * @template D - Source object type(s)
 *
 * @param {T} object - Destination object
 * @param {D} defaults - Object(s) used to assign default properties
 * @return {ExactOptionalPropertyTypes<Merge<T, D> & T>} `object` with defaults
 */
function defu<T extends ObjectPlain, D extends ObjectPlain[]>(
  object: T,
  ...defaults: D
): ExactOptionalPropertyTypes<Merge<T, D> & T> {
  return merge<T, D>(
    /**
     * Determines how to assign a property.
     *
     * @param {unknown} v1 - Original value from {@linkcode object}
     * @param {unknown} v2 - Incoming value from object in {@linkcode defaults}
     * @return {unknown} Property value for {@linkcode object}
     */
    (v1: unknown, v2: unknown): unknown => (isUndefined(v1) ? v2 : v1),
    object,
    ...defaults
  )
}

export default defu
