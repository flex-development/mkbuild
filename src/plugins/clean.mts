/**
 * @file Plugins - clean
 * @module mkbuild/plugins/clean
 */

import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedOutputOptions,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:clean'

/**
 * Cleaned directory cache.
 *
 * @internal
 *
 * @const {Set<string>} directories
 */
const directories: Set<string> = new Set<string>()

plugin.PLUGIN_NAME = PLUGIN_NAME
plugin.directories = directories

export default plugin

/**
 * Create an output directory cleaner plugin.
 *
 * @internal
 *
 * @see {@linkcode Plugin}
 *
 * @this {void}
 *
 * @return {Plugin}
 *  Output directory cleaner plugin
 */
function plugin(this: void): Plugin {
  return {
    name: PLUGIN_NAME,
    renderStart: { handler: renderStart, order: 'pre', sequential: true }
  }

  /**
   * Remove and recreate an output directory.
   *
   * @see https://rollupjs.org/plugin-development/#renderstart
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} output
   *  Normalized output options
   * @return {Promise<undefined>}
   */
  async function renderStart(
    this: PluginContext,
    output: NormalizedOutputOptions
  ): Promise<undefined> {
    ok(typeof output.clean === 'boolean', 'expected `output.clean`')
    ok(typeof output.fs === 'object', 'expected `output.fs`')
    ok(typeof output.root === 'string', 'expected `output.root`')
    ok(typeof output.write === 'boolean', 'expected `output.write`')

    if (
      output.clean &&
      output.root &&
      output.write &&
      typeof output.dir === 'string'
    ) {
      directories.add(output.root)

      /**
       * Absolute path to output directory.
       *
       * @const {string} path
       */
      const path: string = pathe.join(output.root, output.dir)

      if (!directories.has(path)) {
        await output.fs.rm(path, { force: true, recursive: true })
        await output.fs.mkdir(path, { recursive: true })
      }

      directories.add(path)
    }

    return void this
  }
}
