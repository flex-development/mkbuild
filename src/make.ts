/**
 * @file Make
 * @module mkbuild/make
 */

import * as color from 'colorette'
import consola from 'consola'
import { defu } from 'defu'
import type { Format } from 'esbuild'
import fse from 'fs-extra'
import * as pathe from 'pathe'
import type { PackageJson } from 'pkg-types'
import pb from 'pretty-bytes'
import { IGNORE_PATTERNS } from './config/constants'
import loadBuildConfig from './config/load'
import type { Config, Entry, Result } from './interfaces'
import type { EsbuildOptions, OutputExtension } from './types'
import analyzeResults from './utils/analyze-results'
import esbuilder from './utils/esbuilder'
import write from './utils/write'

/**
 * [esbuild][1] based file-to-file transformer and bundler.
 *
 * [1]: https://esbuild.github.io
 *
 * @todo validate build results against package.json
 *
 * @async
 *
 * @example
 *  await make({ entries: [{}] })
 * @example
 *  await make({ entries: [{ ext: '.mjs', format: 'esm' }] })
 * @example
 *  await make({ entries: [{ ext: '.cjs', format: 'cjs' }] })
 * @example
 *  await make({ entries: [{ ext: '.js', format: 'iife' }] })
 *
 * @param {Config} [config={}] - Build configuration options
 * @return {Promise<Result[]>} Build results
 */
async function make({ cwd = '.', ...config }: Config = {}): Promise<Result[]> {
  const {
    bundle,
    clean,
    cwd: root,
    declaration,
    esbuild,
    format,
    fs,
    ignore,
    outdir,
    pattern,
    ...options
  }: Required<Config> = defu(await loadBuildConfig(cwd), config, {
    bundle: false,
    clean: true,
    cwd,
    declaration: true,
    entries: [],
    esbuild: {} as EsbuildOptions,
    ext: '.mjs' as OutputExtension,
    format: 'esm' as Format,
    fs: fse,
    ignore: IGNORE_PATTERNS,
    outdir: 'dist',
    pattern: '**',
    source: 'src'
  })

  // determine current working directory
  cwd = pathe.resolve(process.cwd(), root)

  /**
   * `package.json` data.
   *
   * @var {PackageJson} pkg
   */
  let pkg: PackageJson

  // attempt to read package.json
  try {
    pkg = await fs.readJson(pathe.resolve(cwd, 'package.json'))
  } catch {
    consola.error('package.json not found')
    return []
  }

  /**
   * Normalized build entries.
   *
   * @const {Entry[]} entries
   */
  const entries: Entry[] = options.entries.map(entry => {
    const { peerDependencies = {} } = pkg

    entry.absWorkingDir = cwd
    entry.bundle = entry.bundle ?? bundle
    entry.format = entry.format ?? format
    entry.source = entry.source ?? (entry.bundle ? 'src/index' : 'src')

    return defu(entry, esbuild, {
      bundle: entry.bundle,
      declaration,
      ext: (entry.format === 'esm' ? '.mjs' : '.js') as Entry['ext'],
      external: Object.keys(peerDependencies),
      format,
      ignore,
      name: undefined,
      outbase: entry.bundle ? pathe.parse(entry.source).root : entry.source,
      outdir,
      pattern,
      source: entry.source
    })
  })

  // print build start info
  consola.info(color.cyan(`Building ${pkg.name}`))

  // clean output directories
  if (clean) {
    /**
     * Output directory paths.
     *
     * @const {string[]} outdirs
     */
    const outdirs: string[] = entries.map(entry => {
      return pathe.resolve(cwd, entry.outdir)
    })

    // remove and recreate output directories
    for (const outdir of new Set(outdirs)) {
      await fs.unlink(outdir).catch(() => ({}))
      await fs.emptyDir(outdir)
      await fs.mkdirp(outdir)
    }
  }

  /**
   * Written build results.
   *
   * @const {[string, Result[]][]} written
   */
  const written: [string, Result[]][] = []

  // process build entries
  for (const entry of entries) {
    /**
     * Build results for {@link entry}.
     *
     * @var {Result[]} results
     */
    let results: Result[] = []

    // attempt to build source files
    try {
      results = await esbuilder(entry, fs)
    } catch {
      return []
    }

    // write build results
    for (const result of results) {
      try {
        await write(result, fs)
      } catch (e: unknown) {
        consola.error((e as Error).message)
        return []
      }
    }

    // push written build results
    written.push([entry.outdir, results])
  }

  /**
   * Total build size.
   *
   * @var {number} bytes
   */
  let bytes: number = 0

  // print build done info
  consola.success(color.green(`Build succeeded for ${pkg.name}`))

  // print build analysis
  for (const [outdir, results] of written) {
    bytes += results.reduce((acc, result) => acc + result.bytes, 0)
    consola.log(analyzeResults(outdir, results))
  }

  // print total build size
  consola.log('Î£ Total build size:', color.cyan(pb(bytes)))

  return written.flatMap(([, results]) => results)
}

export default make
