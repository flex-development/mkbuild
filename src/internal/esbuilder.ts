/**
 * @file Internal - esbuilder
 * @module mkbuild/internal/esbuilder
 */

import type { Entry, Result, SourceFile } from '#src/interfaces'
import * as mkp from '#src/plugins'
import type { FileSystemAdapter, OutputMetadata } from '#src/types'
import fsa from '#src/utils/fs'
import IGNORE from '#src/utils/ignore-patterns'
import loaders from '#src/utils/loaders'
import { EXT_DTS_REGEX } from '@flex-development/ext-regex'
import * as mlly from '@flex-development/mlly'
import * as pathe from '@flex-development/pathe'
import * as esbuild from 'esbuild'
import regexp from 'escape-string-regexp'
import fg from 'fast-glob'
import { omit } from 'radash'

/**
 * Builds `entry` using the [esbuild build API][1].
 *
 * [1]: https://esbuild.github.io/api/#build-api
 *
 * @todo ensure `entry.source` is a directory when bundling is disabled
 * @todo ensure `entry.source` is a file (with extname) when bundling is enabled
 *
 * @async
 *
 * @param {Partial<Entry>} entry - Build entry object
 * @param {Partial<FileSystemAdapter>} [fs=fsa] - Custom file system methods
 * @return {Promise<[esbuild.Metafile, Result[]]>} Metafile and build results
 * @throws {esbuild.BuildFailure}
 */
const esbuilder = async (
  entry: Partial<Entry>,
  fs: Partial<FileSystemAdapter> = fsa
): Promise<[esbuild.Metafile, Result[]]> => {
  const {
    bundle = false,
    conditions = bundle ? [] : mlly.CONDITIONS,
    createRequire: insertRequire,
    cwd = '.',
    dts: declaration,
    ext = '.mjs',
    external = [],
    format = 'esm',
    ignore = IGNORE,
    loader = {},
    name = '[name]',
    outExtension = {},
    outdir = 'dist',
    pattern = '**',
    platform = '',
    plugins = [],
    resolveExtensions = mlly.RESOLVE_EXTENSIONS,
    source = bundle ? 'src/index' : 'src',
    tsconfig = '',
    write = false,
    outbase = bundle ? pathe.parse(source).dir : source,
    ...options
  } = entry

  /**
   * Absolute path to current working directory.
   *
   * @const {string} absWorkingDir
   */
  const absWorkingDir: string =
    pathe.resolve(cwd).replace(/\/$/, '') + pathe.sep

  /**
   * Relative paths to source files.
   *
   * **Note**: Files are relative to {@linkcode source} when {@linkcode bundle}
   * is `false`. When enabled, files are relative to {@linkcode absWorkingDir}.
   *
   * @const {string[]} files
   */
  const files: string[] = await fg(bundle ? source : pattern, {
    braceExpansion: true,
    caseSensitiveMatch: true,
    cwd: pathe.join(absWorkingDir, bundle ? '' : source),
    dot: true,
    extglob: true,
    fs: omit(fs, ['readdir']),
    globstar: true,
    ignore: [...new Set(ignore)],
    objectMode: false,
    onlyFiles: true,
    stats: false,
    throwErrorOnBrokenSymbolicLink: false,
    unique: true
  })

  /**
   * Source file objects.
   *
   * @const {SourceFile[]} sources
   */
  const sources: SourceFile[] = files.map(file => {
    /**
     * Absolute path to source file.
     *
     * @const {string} path
     */
    const path: string = pathe.join(absWorkingDir, source, bundle ? '' : file)

    /**
     * File extension of source file.
     *
     * @var {string} ext
     */
    let ext: string = pathe.extname(file)

    // fix file extension if source file is typescript declaration file
    /* c8 ignore next */ EXT_DTS_REGEX.test(file) && (ext = '.d' + ext)

    return {
      ext: ext as pathe.Ext,
      file: bundle
        ? // adds outbase support for bundles (esbuild only uses outbase for
          // multiple entry points); @see https://esbuild.github.io/api/#outbase
          file.replace(new RegExp(`^${regexp(outbase)}`), '').replace(/^\//, '')
        : file,
      path
    }
  })

  // add fully-specified plugin
  if (format === 'esm' || ext === '.cjs') plugins.unshift(mkp.fullySpecified())

  // add tsconfig-paths plugin
  tsconfig && plugins.unshift(mkp.tsconfigPaths())

  // add dts plugin
  if (declaration as boolean) plugins.unshift(mkp.dts())

  // add create-require plugin
  if ((bundle && format === 'esm' && platform === 'node') || insertRequire) {
    plugins.unshift(mkp.createRequire())
  }

  // add decorators plugin
  tsconfig && plugins.unshift(mkp.decorators())

  // add output file writer plugin
  if (write) {
    plugins.push(
      mkp.write({
        ...fs,
        filter: declaration === 'only' ? EXT_DTS_REGEX : undefined
      })
    )
  }

  // remove unsupported options that are not overridden
  Reflect.deleteProperty(options, 'incremental')
  Reflect.deleteProperty(options, 'outfile')
  Reflect.deleteProperty(options, 'stdin')
  Reflect.deleteProperty(options, 'watch')

  // build source files
  const { errors, metafile, outputFiles, warnings } = await esbuild.build({
    ...options,
    absWorkingDir,
    bundle,
    conditions: [...new Set(conditions)],
    entryNames: `[dir]/${name}`,
    entryPoints: sources.reduce<Record<string, string>>((acc, src) => {
      const { ext, file } = src

      /**
       * Regex pattern matching {@linkcode ext}.
       *
       * @const {RegExp} ext_re
       */
      const ext_re: RegExp = new RegExp(`${regexp(ext)}$`)

      acc[file.replace(ext_re, '')] = pathe.join(source, bundle ? '' : file)
      return acc
    }, {}),
    external: bundle ? [...new Set(external)] : [],
    format,
    loader: { ...loaders(format, bundle), ...loader },
    metafile: true,
    outExtension: { ...outExtension, '.js': ext },
    outbase,
    outdir,
    platform: platform as esbuild.Platform,
    plugins,
    resolveExtensions: [...new Set(resolveExtensions)].map(ext => {
      return pathe.formatExt(ext)
    }),
    tsconfig,
    write: false
  })

  return [
    metafile,
    outputFiles
      .map(({ contents, path, text }) => {
        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@linkcode absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = path.replace(
          new RegExp(`^${regexp(absWorkingDir)}`),
          ''
        )

        /**
         * Output metadata.
         *
         * @const {OutputMetadata} metadata
         */
        const metadata: OutputMetadata = metafile.outputs[outfile]!

        return {
          bytes: metadata.bytes,
          contents,
          entryPoint: metadata.entryPoint,
          errors,
          exports: [...new Set(metadata.exports)],
          imports: metadata.imports,
          inputs: metafile.inputs,
          outfile,
          path,
          text,
          warnings
        }
      })
      .filter(res => {
        return declaration === 'only' ? EXT_DTS_REGEX.test(res.path) : true
      })
  ]
}

export default esbuilder
