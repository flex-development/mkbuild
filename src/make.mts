/**
 * @file make
 * @module mkbuild/make
 */

import readPackageJson from '#internal/read-package-json'
import clean from '#plugins/clean'
import mergeTask from '#utils/merge-task'
import runnableTask from '#utils/runnable-task'
import type { Config, Report, RunnableTask } from '@flex-development/mkbuild'

/**
 * Execute build tasks.
 *
 * @see {@linkcode Config}
 * @see {@linkcode Report}
 *
 * @async
 *
 * @this {void}
 *
 * @param {Config | null | undefined} [config]
 *  Build configuration
 * @return {Promise<Report>}
 *  Build report
 */
async function make(
  this: void,
  config?: Config | null | undefined
): Promise<Report> {
  /**
   * Build report.
   *
   * @const {Report} report
   */
  const report: Report = Object.defineProperties({ builds: [], size: 0 }, {
    size: {
      enumerable: true,
      /**
       * Get the size of `this` build report.
       *
       * @this {Report}
       *
       * @return {number}
       *  Build report size in bytes
       */
      get(this: Report): number {
        return this.builds.reduce((acc, result) => acc + result.size, 0)
      }
    }
  })

  if (config) {
    const { fs, tasks, ...baseTask } = config

    // execute build tasks
    for (const tsk of tasks ?? [baseTask]) {
      /**
       * Runnable build task.
       *
       * @const {RunnableTask} task
       */
      const task: RunnableTask = runnableTask(mergeTask({}, baseTask, tsk), fs)

      // run build task + add result to build report
      report.builds.push(await task.run())
    }

    // clear caches
    clean.directories.clear()
    readPackageJson.cache.clear()
  }

  return report
}

export default make
