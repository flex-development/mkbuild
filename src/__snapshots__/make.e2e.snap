// Vitest Snapshot v1

exports[`e2e:make > cjs > should create bundle in cjs format 1`] = `
[
  {
    "entryPoint": "src/index.cts",
    "errors": [],
    "outfile": "dist/index.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/my-atoi/dist/index.cjs",
    "warnings": [],
  },
]
`;

exports[`e2e:make > cjs > should create bundleless build in cjs format 1`] = `
[
  {
    "entryPoint": "src/index.cts",
    "errors": [],
    "outfile": "dist/index.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/my-atoi/dist/index.cjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/max-value.ts",
    "errors": [],
    "outfile": "dist/max-value.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/my-atoi/dist/max-value.cjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/min-value.ts",
    "errors": [],
    "outfile": "dist/min-value.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/my-atoi/dist/min-value.cjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/my-atoi.cts",
    "errors": [],
    "outfile": "dist/my-atoi.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/my-atoi/dist/my-atoi.cjs",
    "warnings": [],
  },
]
`;

exports[`e2e:make > esm > should create bundle in esm format 1`] = `
[
  {
    "entryPoint": "dbl-linear.ts",
    "errors": [],
    "outfile": "dist/index.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/dbl-linear/dist/index.mjs",
    "warnings": [],
  },
]
`;

exports[`e2e:make > esm > should create bundleless build in esm format 1`] = `
[
  {
    "entryPoint": "dbl-linear.ts",
    "errors": [],
    "outfile": "dist/dbl-linear.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/dbl-linear/dist/dbl-linear.mjs",
    "warnings": [],
  },
]
`;

exports[`e2e:make > iife > should create bundle in iife format 1`] = `
[
  {
    "entryPoint": "src/tribonacci.ts",
    "errors": [],
    "outfile": "dist/index.js",
    "path": "\${process.cwd()}/__fixtures__/pkg/tribonacci/dist/index.js",
    "warnings": [],
  },
]
`;

exports[`e2e:make > iife > should create bundleless build in iife format 1`] = `
[
  {
    "entryPoint": "src/tribonacci.ts",
    "errors": [],
    "outfile": "dist/tribonacci.js",
    "path": "\${process.cwd()}/__fixtures__/pkg/tribonacci/dist/tribonacci.js",
    "warnings": [],
  },
]
`;

exports[`e2e:make > should copy files to output directory 1`] = `
[
  {
    "entryPoint": "sum-of-intervals.cjs",
    "errors": [],
    "outfile": "dist/sum-of-intervals.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/sum-of-intervals/dist/sum-of-intervals.cjs",
    "warnings": [],
  },
  {
    "entryPoint": "sum-of-intervals.d.ts",
    "errors": [],
    "outfile": "dist/sum-of-intervals.d.ts",
    "path": "\${process.cwd()}/__fixtures__/pkg/sum-of-intervals/dist/sum-of-intervals.d.ts",
    "warnings": [],
  },
  {
    "entryPoint": "sum-of-intervals.mjs",
    "errors": [],
    "outfile": "dist/sum-of-intervals.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/sum-of-intervals/dist/sum-of-intervals.mjs",
    "warnings": [],
  },
]
`;
